codegen/src/types/substitutes.rs

```diff
@@ -265,14 +401,80 @@ fn is_absolute(path: &syn::Path) -> bool {
 pub struct AbsolutePath(pub syn::Path);
 
 impl TryFrom<syn::Path> for AbsolutePath {
-    type Error = (syn::Path, String);
+    type Error = TypeSubstitutionError;
     fn try_from(value: syn::Path) -> Result<Self, Self::Error> {
         if is_absolute(&value) {
             Ok(AbsolutePath(value))
         } else {
-            Err(
-                (value, "The substitute path must be a global absolute path; try prefixing with `::` or `crate`".to_owned())
-            )
+            Err(TypeSubstitutionError::ExpectedAbsolutePath(value.span()))
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    macro_rules! syn_path {
+        ($path:path) => {{
+            let path: syn::Path = syn::parse_quote!($path);
+            path
+        }};
+    }
+
+    macro_rules! type_path {
+        ($path:path) => {{
+            let path: syn::Path = syn::parse_quote!($path);
+            TypePath::from_syn_path(path)
+        }};
+    }
+
+    fn ident(name: &'static str) -> syn::Ident {
+        syn::Ident::new(name, proc_macro2::Span::call_site())
+    }
+
+    #[test]
+    fn replacing_nested_type_params_works() {
+        // Original path, replacement ident->paths, expected output path
+        let paths = [
+            (
+                syn_path!(::some::path::Foo<::other::Path<A, B>>),
+                vec![],
+                syn_path!(::some::path::Foo<::other::Path<A, B>>),
+            ),
+            (
+                syn_path!(::some::path::Foo<::other::Path<A, B>>),
+                vec![(ident("A"), type_path!(::new::Value))],
+                syn_path!(::some::path::Foo<::other::Path<::new::Value, B>>),
+            ),
+            (
+                syn_path!(::some::path::Foo<::other::Path<A, B>>),
+                vec![
+                    (ident("A"), type_path!(::new::A)),
+                    (ident("B"), type_path!(::new::B)),
+                ],
+                syn_path!(::some::path::Foo<::other::Path<::new::A, ::new::B>>),
+            ),
+            (
+                syn_path!(
+                    ::some::path::Foo<::other::Path<A, ::more::path::to<::something::Argh<B>>>, C>
```

Comment:
nit: Would it also work for `  ::some::path::Foo<::other::Path<A, ::more::path::to<::something::Argh<B>>>, C, A>` where we have 2 `A` to replace with `::new::A`?

Reply:
Good idea for a test; it should work like that; any same named ident will get replaced. Lemme add a test :)